File: src/main/java/me/testaccount666/serversystem/moderation/AbstractModerationManager.java
package me.testaccount666.serversystem.moderation;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public abstract class AbstractModerationManager {
    protected UUID ownerUuid;

    public AbstractModerationManager(UUID ownerUuid) {
        this.ownerUuid = ownerUuid;
    }

    public abstract void addModeration(AbstractModeration moderation);

    public abstract void removeModeration(AbstractModeration moderation);

    public abstract List<AbstractModeration> getModerations();

    public boolean hasActiveModeration() {
        return getModerations().stream().anyMatch(moderation -> !moderation.isExpired());
    }

    public Optional<AbstractModeration> getActiveModeration() {
        return getModerations().stream().filter(moderation -> !moderation.isExpired()).findFirst();
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/moderation/AbstractModerationManager.kt
package me.testaccount666.serversystem.moderation

import me.testaccount666.serversystem.ServerSystem.Companion.instance
import me.testaccount666.serversystem.managers.database.moderation.ModerationDatabaseManager
import java.util.*

abstract class AbstractModerationManager<T : AbstractModeration>(val ownerUuid: UUID) {
    protected val databaseManager = instance.registry.getService<ModerationDatabaseManager>()

    abstract fun addModeration(moderation: T)

    abstract fun removeModeration(moderation: T)

    abstract val moderations: List<T>

    open val activeModeration
        get() = moderations.firstOrNull { !it.isExpired }

    open val activeModerations
        get() = moderations.filter { !it.isExpired }

    fun hasActiveModeration() = moderations.any { !it.isExpired }
}
